packages/hardhat-ledger/src/provider.ts

```diff
@@ -0,0 +1,445 @@
+
+    await this._requireControlledInit(txRequest.from);
+
+    if (txRequest.gas === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_TX_PARAM_TO_SIGN_LOCALLY, {
+        param: "gas",
+      });
+    }
+
+    if (txRequest.from === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_TX_PARAM_TO_SIGN_LOCALLY, {
+        param: "from",
+      });
+    }
+
+    const hasGasPrice = txRequest.gasPrice !== undefined;
+    const hasEip1559Fields =
+      txRequest.maxFeePerGas !== undefined ||
+      txRequest.maxPriorityFeePerGas !== undefined;
+
+    if (!hasGasPrice && !hasEip1559Fields) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_FEE_PRICE_FIELDS);
+    }
+
+    if (hasGasPrice && hasEip1559Fields) {
+      throw new HardhatError(ERRORS.NETWORK.INCOMPATIBLE_FEE_PRICE_FIELDS);
+    }
+
+    if (hasEip1559Fields && txRequest.maxFeePerGas === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_TX_PARAM_TO_SIGN_LOCALLY, {
+        param: "maxFeePerGas",
+      });
+    }
+
+    if (hasEip1559Fields && txRequest.maxPriorityFeePerGas === undefined) {
+      throw new HardhatError(ERRORS.NETWORK.MISSING_TX_PARAM_TO_SIGN_LOCALLY, {
+        param: "maxPriorityFeePerGas",
+      });
+    }
+
+    const path = await this._derivePath(txRequest.from);
+
+    if (txRequest.nonce === undefined) {
+      txRequest.nonce = await this._getNonce(txRequest.from);
+    }
+
+    const chainId = await this._getChainId();
+
+    const baseTx: ethers.TransactionLike = {
+      chainId,
+      data:
+        txRequest.data === undefined ? undefined : this._toHex(txRequest.data),
+      gasLimit: txRequest.gas,
+      gasPrice: txRequest.gasPrice,
+      maxFeePerGas: txRequest.maxFeePerGas,
+      maxPriorityFeePerGas: txRequest.maxPriorityFeePerGas,
+      nonce: Number(txRequest.nonce),
+      value: txRequest.value,
+    };
+    if (txRequest.to !== undefined) {
+      baseTx.to = this._toHex(txRequest.to);
+    }
+
+    const txToSign =
+      ethers.Transaction.from(baseTx).unsignedSerialized.substring(2);
+
+    const resolution = await ledgerService.resolveTransaction(txToSign, {}, {});
+
+    const signature = await this._withConfirmation(() =>
+      this.eth.signTransaction(path, txToSign, resolution)
+    );
+
+    const rawTransaction = ethers.Transaction.from({
+      ...baseTx,
+      signature: {
+        v: this._toHex(signature.v),
+        r: this._toHex(signature.r),
+        s: this._toHex(signature.s),
+      },
+    }).serialized;
+
+    return this._wrappedProvider.request({
+      method: "eth_sendRawTransaction",
+      params: [rawTransaction],
+    });
+  }
+
+  private async _derivePath(addressToFindAsBuffer: Buffer): Promise<string> {
+    const addressToFind = this._toHex(addressToFindAsBuffer).toLowerCase();
+
+    if (this.paths[addressToFind] !== undefined) {
+      return this.paths[addressToFind];
+    }
+
+    this.emit("derivation_start");
+
+    let path = "";
```

Comment:
When I saw this, I tried to convert it into a `string | undefined` variable, but I see that that makes the code unnecessarily more complex, because we "know" this variable is going to be immediately set.

That being said, in these cases I prefer to use a more recognizable value, in case this ends up being shown in an error message for any reason. So what do you think about something like:

```suggestion
    let path = "<unset-path>";
```

Reply:
I've never thought of that but I like it! It can also act as a failsafe if the variable is not properly set later (and there are tests)

The wonders if having to move variables out of scope to be able to use them in errors haha, I'll change it and keep it in mind for future cases